4단계 - 계산기 UI

devsoupe opened the pull request:
에뮬레이터에서 Espresso 테스트시 문제가 없는데 실제 단말에서 테스트시(플립3)
androidx.test.espresso.PerformException: Error performing 'single click' on view 'with id is <edu.nextstep.camp.calculator:id/buttonPlus>' 이런 에러가 나면서 테스트 케이스를 실패하네요.

뷰 Visibility 문제 인것 같긴한데... 설정에서 애니메이션 관련도 다 disable 해도 동일해서 관련하여 혹시 조언해주실만 내용이 있으신지도 리뷰와 함께 문의드립니다.

devsoupe committed:
feed : 코드리뷰 요청사항 수정

devsoupe committed:
feat : README.md 업데이트

devsoupe committed:
feat : Expression, ExpressionGenerator 구현

devsoupe committed:
test : 기능 요구 사항 ExpressionGeneratorTest 작성

devsoupe committed:
refactor : delimiter 책임 변경, 수식 문자 변경

devsoupe committed:
feat : 계산기 UI 작성 및 Calculator 기능 추가

devsoupe committed:
test : CalculatorTest 추가 및 변경

devsoupe committed:
refactor : Calculator 클래스 변경, 결과 응답, 인덴트

devsoupe committed:
test : 기능 요구 사항 UI MainActitityTest 작성

devsoupe committed:
test : 누락된 기능 테스트 작성 (실패)

devsoupe committed:
test : 누락된 기능 도메인 테스트 작성 (실패)

devsoupe committed:
feat : 누락된 기능 구현

devsoupe committed:
test : 기능 요구 사항 UI MainActivityTest 추가 작성 및 수정

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/BinaryCalculator.kt:
```diff
@@ -15,8 +15,8 @@ internal class BinaryCalculator private constructor(private val operator: Operat
             "-" -> { x, y -> x - y }
-            "*" -> { x, y -> x * y }
-            "/" -> { x, y -> x / y }
+            "×" -> { x, y -> x * y }
+            "÷" -> { x, y -> x / y }
```
UI에서의 제약사항으로 인해 `*` --> `x`  `/` --> `÷`로 변경되는것에 대해 어떻게 생각하시나요?
만약 후에 곱하기와 나누기를 다시 원래 기호로 되돌리거나 다른 기호가 되었을 때도 매번 도메인 로직이 변경되야 하지는 않을까요?

UI에서의 기능을 적절한 형태로 가공해서 도메인의 입력값으로 만들어 주거나
곱하기 기호에 두개를 사용하는것도 한가지 확장 방법일 수 있을 듯 합니다~

devsoupe replied:
도메인에서 Operator를 제공해주면 더 좋을것 같긴한데, 우선 심플하게 UI에서 매직문자로 도메인 입력값으로 변경하였습니다. ㅎㅎ

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -1,13 +1,52 @@
-        val chunks = ExpressionParser.parse(expression, delimiter)
+class Calculator {
+
+    sealed class Result {
+
+        data class Success(val value: String): Result()
+        object Failure: Result()
+    }
```
결과를 sealed class로 만들어 여러가지 상황에 적절하게 대응할 수 있게 만들어 주신 점 좋습니다. 👍 

다만 Failure의 경우 어떤 원인으로 실패하였는지 알 방법이 없는데요.
sealed class의 장점을 이용해 실패 케이스를 확장하는 형태도 괜찮을 것 같습니다.

```kotlin
sealed class Result {
    object xxxFailure
    object yyyFailure
    ...
```

혹은 Failure 타입을 유지한다면, cause를 알 수 있게 만들어 주시면 좋겠네요!
이 경우는 Kotlin.Result 타입을 사용해서 간단하게 처리가 가능합니다

devsoupe replied:
Exception추가와 더불어 kotlin.Result를 활용해 보겠습니다!

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -1,13 +1,52 @@
+        object Failure: Result()
+    }
+
+    private val expressionGenerator = ExpressionGenerator(DELIMITER)
+    private val expressionParser = ExpressionParser(DELIMITER)
```
DELIMITER는 UI에 따라 변경될 수 있기 때문에 입력을 받는 형태의 구조는 유지하는게 좋을 듯 해요!

devsoupe replied:
이전 단계 코드리뷰에서 말씀해주셔서 다시 바꾸면서 고민했던 부분인데요. 
기존에는 구분자가 UI(외부)에서 결정되어진다고 생각을 했었는데(수식결정이 UI쪽), 이번에 ExpressionGenerator를 만들면서
UI(외부)에서는 구분자를 결정하는 책임이 사라졌고, 숫자/연산자 심볼만 전달해주기에 각 심볼을 어떠한 구분자로 묶어서
expression을 만드는지는 도메인쪽으로 책임이 넘어왔다고 생각을 했거든요. 그래서 Calculator로 다시 변경을 하였습니다.
이에 대해 의견여쭤볼게요. ㅎㅎ 그리고 현재 구조에서도 "DELIMITER는 UI에 따라 변경될수 있다"라고 한다면 이에 대한 예제를 부탁드려봐도 될까요? 제가 정확히 이해를 잘 못하고 있는것 일수도 있어서요!

laco-dev replied:
말씀 잘 정리해주셔서 이해가 된 것 같습니다.
계산기의 경우 입력 값으로 하나의 문자를 받게 되는 형태이니 도메인 기능으로 역할이 위임되었다고 볼 수 있을 것 같습니다.

만약 Calculator가 expression을 받는 구조라면 delimiter를 같이 받는 것을 고려해 볼 수 있을 것 같은데 symbol 형태의 구조에서는 지금 구조가 더 나아 보이네요! 👍

devsoupe replied:
말씀해주신 DELIMITER로 객체의 책임에 대해 많은 생각을 하게된 것 같습니다. 좋은 인사이트를 주셔서 감사합니다!

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -1,13 +1,52 @@
+            return
+        }
+
+        val expression = expressionGenerator.generate()
+        val evaluated = evaluate(expression)
+
+        expressionGenerator.update(evaluated)
+        result = Result.Success(evaluated)
```
evaluated 를 계산하는 과정에서도 충분히 오류가 발생할 수 있습니다.
이 경우는 런타임 예외로 앱이 종료되는 문제로 직결될 수 있는데, Failure 타입을 만든것과 유사하게
적절한 예외 타입으로 만들어 줄 필요가 있겠습니다.

kotlin.runCatching { ... } 함수 참고해 주세요

devsoupe replied:
넵. runCatching로 evaluated 에러도 예외처리 하도록 하겠습니다!

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/ExpressionGenerator.kt:
```diff
@@ -0,0 +1,51 @@
+
+    private val expression = Expression()
+
+    fun append(symbol: String) = when {
+        isDigit(symbol) -> processDigit(symbol)
+        else -> processOperator(symbol)
+    }
```
```kotlin
fun append(symbol: String) = apply { 
    when ( ... }   
}
```
형태로 만들면 processXXX 함수에 불필요한 코드를 포함시킬 필요가 없을 것 같습니다

devsoupe replied:
안그래도 눈엣가시였는데 멋진 활용법이네요. 👍 감사합니다!!

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Expression.kt:
```diff
@@ -0,0 +1,39 @@
+
+internal class Expression {
+
+    private val symbols = mutableListOf<String>()
```
현재까지의 중간 값은 Calculator에서 관리가 되고 있기 때문에 이 곳에서도 모든 중간 값을 가지고 있을 필요는 없을 것 같습니다!

최대한 변경 가능한 상태는 적게 유지하는게 좋습니다.
이로 인해 불필요한 public 함수가 많이 늘어난 점도 있구요

laco-dev replied:
혹은 계산기가 중간 과정을 가지지 않게 된다면 Expression이 이 역할을 할 수 있을 것 같네요

devsoupe replied:
Calculator의 중간값을 제거하는 형태로 진행하겠습니다!

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -1,13 +1,52 @@
+    private val expressionGenerator = ExpressionGenerator(DELIMITER)
+    private val expressionParser = ExpressionParser(DELIMITER)
+
+    var result: Result? = null
+        private set
```
중간 계산 과정 때문에 Calculator의 도메인 로직이 많이 변경이 된 것 같습니다.
기존 구현을 유지하고, 반환 타입을 Result로 만드는 형태로 개선해 보는건 어떨까요?

다음 입력 값이 평가되는 시기는 "=" 버튼을 누를 때이니, 계산기의 역할을 다른곳으로 옮길 수 있을 것 같아요.

도메인 로직의 경우 가급적 변경이 불가능한 순수 함수 형태로 만드는게 권장드리고 싶지만, 이를 위한 구현 복잡도가 또 올라가는 문제도 있어서 적절한 도전 과제로 생각 해보셔도 됩니다

devsoupe replied:
1. 상태값은 제거하고 순수함수 형태로
2. Calculator 역할은 원래대로 계산 역할에만 집중
3. 외부와의 인터페이스 역할은 다른 클래스로 위임

이렇게 진행해보도록 하겠습니다~

laco-dev commented on app/src/androidTest/java/edu/nextstep/camp/calculator/MainActivityTest.kt:
```diff
@@ -106,4 +106,122 @@ class MainActivityTest {
+    @Test
+    fun buttonPressOperator_noOperand() {
+        // when : -> +, -, ×, ÷ 클릭 ->
+        onView(withId(R.id.buttonPlus)).perform(click())
+        onView(withId(R.id.buttonMinus)).perform(click())
+        onView(withId(R.id.buttonMultiply)).perform(click())
+        onView(withId(R.id.buttonDivide)).perform(click())
```
테스트 함수를 보다 가독성 있게 만드는것도 한가지 좋은 방법이 될 수 있습니다.
반복적인 View Clicks 처리를 위한 함수를 만들고 아래와 같이 사용을 할 수도 있습니다.
```kotlin
clicks(R.id.buttonPlus, R.id.buttonMinus, R.id.buttonMultiply, R.id.buttonDivide)
```

이건 한가지 예시라, 보다 편한 방법이 있으시면 사용하셔도 됩니다~

devsoupe replied:
말씀해주신대로 가독성과 반복처리 효율을 높일 수 있을것 같습니다. 
우선 특정 테스트에 활용해 보겠습니다~

laco-dev commented on app/src/main/java/edu/nextstep/camp/calculator/MainActivity.kt:
```diff
@@ -17,13 +23,37 @@ class MainActivity : AppCompatActivity() {
+        }
+    }
+
+    private fun setupButtonsClickListener(button: Button) {
+        button.setOnClickListener {
+            calculator.insert(button.text.toString())
+            updateDisplay()
+        }
+    }
```
delete(), evaluate(), insert() 요청에 대한 결과 이후 result가 정말 정확하다고 신뢰할 수 있을까요?
가령, 위 함수들의 동작이 여러 개의 쓰레드에서 각각 다른 연산 속도를 가진다면, 결과가 어떻게 될까요?

이번 과제에서는 이 문제가 직접적으로 드러나지는 않는데 UI와 비즈니스 로직의 구현 관계에서는 생각해볼만한 고민거리라, 남겨보았습니다. 🤔

devsoupe replied:
Expression이 token들 저장하고 있다보니 말씀해주신 동기화 이슈가 생길것 같네요.
Expression을 참조하는 ExpressionGenerator 에서 delete, evaluate, insert 호출시 synchronized로 generator를 감싸거나
비슷한 방법을 Expression 함수에서 할 수 있을 것 같긴한데...
synchonized의 성능 이슈로 순수함수 형태로는 해결방법이 없을까? 하는 생각이 듭니다. ㅎㅎ
함수형에 익숙치가 않아서 혹시라도 관련하여 힌트나 참고해볼만한 내용이 있으시면 공유해주시면 감사하겠습니다!
저 또한 좀더 고민해보겠습니다.

laco-dev reviewed (changes requested):
4단계 미션 수고하셨습니다.
난이도가 꽤 있는데에도 잘 해주셨네요. 👍 

현재의 구조가 다음 미션을 거쳐가면서 어떻게 더욱 개선될지 기대가 됩니다.
고민해볼 만한 내용과 피드백을 코멘트로 남겼으니 반영 후 다시 리뷰 요청 주세요~

laco-dev commented:
> 에뮬레이터에서 Espresso 테스트시 문제가 없는데 실제 단말에서 테스트시(플립3)
androidx.test.espresso.PerformException: Error performing 'single click' on view 'with id is edu.nextstep.camp.calculator:id/buttonPlus' 이런 에러가 나면서 테스트 케이스를 실패하네요.

음.. 🤔 
직접적으로 겪어보지 않아, 정확한 원인은 모르겠는데 제 추측으로는 Activity가 만들어지는 과정(애니메이션 혹은 Cold Start)에 클릭 이벤트를 너무 빠르게 수신받아 뷰를 찾을 수 없는 문제가 아닐까.. 생각해봅니다

devsoupe committed:
feed : apply scope 함수를 활용해 불필요한 코드 제거

devsoupe committed:
feed : 테스트 함수 가독성을 위한 반복처리 함수 리펙토링 (특정 부분에만 우선 활용)

devsoupe committed:
feed : kotiln.Result<T> 사용으로 변경

devsoupe committed:
feed :

1. Calculator 상태값은 제거하고 순수함수 형태로
2. Calculator 역할은 원래대로 계산 역할에만 집중
3. 외부와의 인터페이스 역할은 CalculatorInterface로 위임
4. 역할 분리로 인한 테스트 코드 분리
5. 추가 리팩토링

devsoupe committed:
feed : runCatching를 활용해 evaluate 함수도 예외처리

devsoupe committed:
feed : 연산자 입력값 가공한 형태로 도메인쪽으로 전달, 패키지 리펙토링

laco-dev commented on app/src/androidTest/java/edu/nextstep/camp/calculator/AndroidTestHelper.kt:
```diff
@@ -0,0 +1,21 @@
+
+    fun onViewMatchText(@IdRes id: Int, text: String) {
+        onView(withId(id)).check(matches(withText(text)))
+    }
```
테스트 유틸 👍

laco-dev commented on app/src/main/java/edu/nextstep/camp/calculator/MainActivity.kt:
```diff
@@ -17,13 +23,72 @@ class MainActivity : AppCompatActivity() {
+        when (result.isSuccess) {
+            true -> updateDisplayOnSuccess(result.getOrNull())
+            false -> updateDisplayOnFailure(result.exceptionOrNull())
+        }
```
```kotlin
result
    .onSuccess { ... }
    .onFailure { ... } 
```
위 형태로 성공, 실패에 대한 처리가 가능합니다!

devsoupe replied:
when 절 없이 깔끔하네요!

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/CalculatorInterface.kt:
```diff
@@ -0,0 +1,38 @@
+            calculator.evaluate(expression)
+        }.onSuccess {
+            Result.success(expressionGenerator.update(it).generate())
+        }.onFailure {
+            Result.failure<Throwable>(it)
+        }
```
runCatching의 기본 동작이 오류가 발생하였을 때 Result.failure(throwable)을 만들어 주기 때문에 불필요합니다~

devsoupe replied:
넵!

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/CalculatorInterface.kt:
```diff
@@ -0,0 +1,38 @@
+            val expression = expressionGenerator.generate()
+            calculator.evaluate(expression)
+        }.onSuccess {
+            Result.success(expressionGenerator.update(it).generate())
```
onSuccess 는 아무런 반환값이 없어서 Result로 새로 만들지 않아도 됩니다.

명시적으로 runCatching 블록 안에서 같이 처리해도 될 것 같아요.
```kotlin
runCatching {
   val result = calculator.evaluate(...)
   expressionGenerator.update(result).generate()
}
```

devsoupe replied:
runCatching 함수 내부 잘 확인해보았습니다! 감사합니다.

laco-dev commented on app/src/main/java/edu/nextstep/camp/calculator/MainActivity.kt:
```diff
@@ -17,13 +23,72 @@ class MainActivity : AppCompatActivity() {
+        binding.tvResultDisplay.text = value ?: return
+    }
+
+    private fun updateDisplayOnFailure(throwable: Throwable?) {
+        when (throwable) {
+            is InvalidExpressionException -> Toast.makeText(this, getString(R.string.error_invalid_expression), Toast.LENGTH_SHORT).show()
+            is EvaluateExpressionException -> Toast.makeText(this, getString(R.string.error_evaluate_expression), Toast.LENGTH_SHORT).show()
```
명시적인 예외 케이스를 정의해 처리 👍

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/expression/Expression.kt:
```diff
@@ -0,0 +1,39 @@
+
+internal class Expression {
+
+    private val symbols = mutableListOf<String>()
```
순수함수로 만드는 법은 객체 내부에 변경 가능한 상태를 없애는 데에서 시작하는 전략이 좋습니다.
Expression이 변경되지 않는 상태를 가지고 있는 구조라면 보다 이펙트가 없는 형태로 작성이 가능할 수도 있을 것 같네요.

symbols을 관리하는 관점에서는 [일급 컬렉션](https://jojoldu.tistory.com/412) 형태로 만들면 더욱 좋을 듯
```kotlin
class Expression(private val symbols: List<String>) {
  fun add(symbol: String): Expression { ... }
  fun removeLast(): Expression { ... }
}

private var expression = Expression("")
fun add(symbol: String) {
   expression = expression.add(symbol)
}
```

현재는 이러한 방법도 있다로 봐 주시고, 반대로 코드 복잡도가 올라가는 점도 있어서 다음에 기회가 되면 한번 도전 해 보세요

devsoupe replied:
이번 커밋후에 밑의 내용과 더불어 한번 리펙토링을 진행해보겠습니다!

devsoupe replied:
추가로 Expression을 상태가 없게 만들더라도 ExpressionGenerator에서 Expression를 상태로 가지니 이 또한 문제가 되는거겠죠? (expression = expression.add(symbol)) 추후에 변경에 참고를 해보려고 문의드려요 ㅎ

laco-dev replied:
어디선가는 변경 가능한 상태를 들고있어야 하는데 실제 앱 환경인 통합 테스트 단위에서 보통 변경하도록 하는 편입니다.
MVC 구조에서 Controller의 역할이 변경 가능한 상태를 들고있습니다. (State Holder라고 부르기도 합니다)

이번 미션에서는 MainActivity가 컨트롤러의 역할을 하니 Calculator와 Expression을 가지고 있겠죠?

devsoupe replied:
넵. 충분히 고민해보겠습니다! 😄

laco-dev commented on domain/src/main/java/edu/nextstep/camp/calculator/domain/expression/ExpressionGenerator.kt:
```diff
@@ -0,0 +1,49 @@
+package edu.nextstep.camp.calculator.domain.expression
+
+internal class ExpressionGenerator(private val joinDelimiter: String) {
```
무언가 느낌으로는
ExpressionGenerator와 Expression이 완전히 동일한 기능을 하는 것으로 생각이 들기도 하네요 🤔

devsoupe replied:
말씀해주신걸 저도 느끼고 있네요 ㅎㅎ 처음에 원래 하나 였는데 ExpressionGenerator의 복잡도가 올라가는거 같아 분리를 했습니다. Expression 순수함수 만드는것과 함께 고민을 좀 더 해보도록 하겠습니다!

laco-dev reviewed (changes requested):
피드백 반영 잘 해주셨습니다. 👍 💯 

마지막 미션이기 때문에 마이너한 코드 수정을 추가로 요청드리니, 확인 부탁드립니다.
(구조 변경은 참고만 해 주시면 됩니다)

devsoupe committed:
feed : kotlin.Result 처리 when 절 대신 onSuccess, onFailure 사용

devsoupe committed:
feed : runCatching 기본 기능을 활용하여 심플하게

devsoupe committed:
refactor : 함수명 리펙토링

laco-dev commented on app/src/main/java/edu/nextstep/camp/calculator/MainActivity.kt:
```diff
@@ -17,13 +23,71 @@ class MainActivity : AppCompatActivity() {
+    }
+
+    private fun updateDisplay(result: Result<String>) {
+        result
+            .onSuccess { updateDisplayOnSuccess(result.getOrNull()) }
+            .onFailure { updateDisplayOnFailure(result.exceptionOrNull()) }
+    }
```
onSuccess, onFailure의 it로 넘어오는 값을 사용하면 됩니다!~

devsoupe replied:
앗 그러네요... ㅋ

laco-dev reviewed (approved):
피드백 반영 확인 헀습니다. 👍 

구조 변경은 충분히 고민 해보고, 다음 2주차 미션에서 같이 진행 해 보셔도 됩니다.
현재의 복잡한 컨트롤러에서 모델을 분리하는 작업을 하게 될 것이기 때문에 보다 접근이 쉬울 수 있어요.

커멘트는 확인만 해 주시면 됩니다. 
1주차 미션 수고 많으셨습니다. 👍
